client/src/main/java/io/atomix/copycat/client/session/ClientSession.java

```diff
@@ -580,12 +589,30 @@ private ClientSession resetConnection() {
    */
   private ClientSession resetMembers() {
     if (connectMembers.isEmpty() || connectMembers.size() < members.size() - 1) {
-      connectMembers = connectionStrategy.getConnections(leader, new ArrayList<>(members));
+      connectMembers = new ArrayList<>(members);
     }
     return this;
   }
 
   /**
+   * Selects the next member to which to send a request.
+   */
+  private Address selectMember() {
+    // Remove the next member that meets the strategy from the members list.
+    Iterator<Address> iterator = connectMembers.iterator();
+    while (iterator.hasNext()) {
+      Address member = iterator.next();
+      if (connectionStrategy.connect(member, leader)) {
+        iterator.remove();
```

Comment:
I think the idea for Atomix is to hide as much of the decision making as possible, but in some cases that may not be possible. For the `RecoveryStrategy` I submitted the PR to handle recovery of all resources in Atomix. I think the idea there is to just recover the resources and then expose an event interface to allow users to listen for events like session open/close/disconnected/etc (which haven't been added yet). For the connection related strategies, I suppose for simplicity we could just allow them to be set on `AtomixClient` for now.